pinot-tools/src/main/java/org/apache/pinot/tools/perf/QueryRunner.java

```diff
@@ -416,42 +422,49 @@ public static QuerySummary multiThreadedQueryRunner(PerfBenchmarkDriverConf conf
+          throw new TimeoutException("Timeout of " + timeout + " sec reached. Aborting");
         }
 
-        long currentTime = System.currentTimeMillis();
-        if (currentTime - reportStartTime >= reportIntervalMs) {
-          long timePassed = currentTime - startTime;
-          int numQueriesExecutedInt = numQueriesExecuted.get();
-          LOGGER.info("Time Passed: {}ms, Queries Executed: {}, Exceptions: {}, Average QPS: {}, "
-                  + "Average Broker Time: {}ms, Average Client Time: {}ms.", timePassed, numQueriesExecutedInt,
-              numExceptions.get(), numQueriesExecutedInt / ((double) timePassed / MILLIS_PER_SECOND),
-              totalBrokerTime.get() / (double) numQueriesExecutedInt,
-              totalClientTime.get() / (double) numQueriesExecutedInt);
-          reportStartTime = currentTime;
-          numReportIntervals++;
-
-          if ((numIntervalsToReportAndClearStatistics != 0) && (numReportIntervals
-              == numIntervalsToReportAndClearStatistics)) {
-            numReportIntervals = 0;
-            startTime = currentTime;
-            reportAndClearStatistics(numQueriesExecuted, numExceptions, totalBrokerTime, totalClientTime,
-                statisticsList);
+        for (String query : queries) {
+          if (timeout > 0 && System.currentTimeMillis() - startTimeAbsolute > timeout) {
+            throw new TimeoutException("Timeout of " + timeout + " sec reached. Aborting");
+          }
+          while (!queryQueue.offer(query)) {
+            Thread.sleep(1);
+          }
+
+          long currentTime = System.currentTimeMillis();
+          if (currentTime - reportStartTime >= reportIntervalMs) {
+            long timePassed = currentTime - startTime;
+            int numQueriesExecutedInt = numQueriesExecuted.get();
+            LOGGER.info("Time Passed: {}ms, Queries Executed: {}, Exceptions: {}, Average QPS: {}, " + "Average "
+                    + "Broker Time: {}ms, Average Client Time: {}ms.", timePassed, numQueriesExecutedInt,
+                numExceptions.get(), numQueriesExecutedInt / ((double) timePassed / MILLIS_PER_SECOND),
+                totalBrokerTime.get() / (double) numQueriesExecutedInt,
+                totalClientTime.get() / (double) numQueriesExecutedInt);
+            reportStartTime = currentTime;
+            numReportIntervals++;
+
+            if ((numIntervalsToReportAndClearStatistics != 0) && (numReportIntervals
+                == numIntervalsToReportAndClearStatistics)) {
+              numReportIntervals = 0;
+              startTime = currentTime;
+              reportAndClearStatistics(numQueriesExecuted, numExceptions, totalBrokerTime, totalClientTime,
+                  statisticsList);
+            }
           }
         }
+        numTimesExecuted++;
       }
-      numTimesExecuted++;
+    } catch (TimeoutException e) {
+      LOGGER.info("Time out reached. {}", e.getMessage());
```

Comment:
timeouts should not be omitted from the reported statistics

Reply:
It's not. Event after this change, you'll still get the message `Time out reached. Timeout of 300000 sec reached. Aborting`
Prior to this change, after this an exception was thrown and no final statistics were printed.
After this change, we stop gracefully ad print the final stats.